// devextreme/esm/renovation/ui/grids/data_grid_next/views/footer.js
/**
 * DevExtreme (esm/renovation/ui/grids/data_grid_next/views/footer.js)
 * Version: 22.1.6
 * Build date: Tue Oct 18 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import _objectDestructuringEmpty from "@babel/runtime/helpers/esm/objectDestructuringEmpty";
import _extends from "@babel/runtime/helpers/esm/extends";
import {
    createComponentVNode
} from "inferno";
import {
    BaseInfernoComponent
} from "@devextreme/runtime/inferno";
import {
    createPlaceholder
} from "../../../../utils/plugin/context";
import {
    Placeholder
} from "../../../../utils/plugin/placeholder";
export var FooterPlaceholder = createPlaceholder();
export var viewFunction = () => createComponentVNode(2, Placeholder, {
    type: FooterPlaceholder
});
export var FooterProps = {};
export class Footer extends BaseInfernoComponent {
    constructor(props) {
        super(props);
        this.state = {}
    }
    get restAttributes() {
        var _this$props = this.props,
            restProps = _extends({}, (_objectDestructuringEmpty(_this$props), _this$props));
        return restProps
    }
    render() {
        this.props;
        return viewFunction()
    }
}
Footer.defaultProps = FooterProps;
