// devextreme/renovation/ui/grids/data_grid_next/scrolling/scrolling.js
/**
 * DevExtreme (renovation/ui/grids/data_grid_next/scrolling/scrolling.js)
 * Version: 22.1.6
 * Build date: Tue Oct 18 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.viewFunction = exports.DataGridNextScrollingProps = exports.DataGridNextScrolling = void 0;
var _inferno = require("inferno");
var _inferno2 = require("@devextreme/runtime/inferno");
var _value_setter = require("../../../../utils/plugin/value_setter");
var _virtual_scrolling = require("./virtual_scrolling");
var _plugins = require("./plugins");
var _excluded = ["mode"];

function _objectWithoutProperties(source, excluded) {
    if (null == source) {
        return {}
    }
    var target = _objectWithoutPropertiesLoose(source, excluded);
    var key, i;
    if (Object.getOwnPropertySymbols) {
        var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
        for (i = 0; i < sourceSymbolKeys.length; i++) {
            key = sourceSymbolKeys[i];
            if (excluded.indexOf(key) >= 0) {
                continue
            }
            if (!Object.prototype.propertyIsEnumerable.call(source, key)) {
                continue
            }
            target[key] = source[key]
        }
    }
    return target
}

function _objectWithoutPropertiesLoose(source, excluded) {
    if (null == source) {
        return {}
    }
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;
    for (i = 0; i < sourceKeys.length; i++) {
        key = sourceKeys[i];
        if (excluded.indexOf(key) >= 0) {
            continue
        }
        target[key] = source[key]
    }
    return target
}

function _extends() {
    _extends = Object.assign ? Object.assign.bind() : function(target) {
        for (var i = 1; i < arguments.length; i++) {
            var source = arguments[i];
            for (var key in source) {
                if (Object.prototype.hasOwnProperty.call(source, key)) {
                    target[key] = source[key]
                }
            }
        }
        return target
    };
    return _extends.apply(this, arguments)
}

function _defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) {
            descriptor.writable = true
        }
        Object.defineProperty(target, descriptor.key, descriptor)
    }
}

function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) {
        _defineProperties(Constructor.prototype, protoProps)
    }
    if (staticProps) {
        _defineProperties(Constructor, staticProps)
    }
    Object.defineProperty(Constructor, "prototype", {
        writable: false
    });
    return Constructor
}

function _inheritsLoose(subClass, superClass) {
    subClass.prototype = Object.create(superClass.prototype);
    subClass.prototype.constructor = subClass;
    _setPrototypeOf(subClass, superClass)
}

function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(o, p) {
        o.__proto__ = p;
        return o
    };
    return _setPrototypeOf(o, p)
}
var viewFunction = function(_ref) {
    var isVirtualScrolling = _ref.isVirtualScrolling,
        mode = _ref.props.mode,
        virtualScrollingMode = _ref.virtualScrollingMode;
    return (0, _inferno.createFragment)([(0, _inferno.createComponentVNode)(2, _value_setter.ValueSetter, {
        type: _plugins.ScrollingModeValue,
        value: mode
    }), isVirtualScrolling && (0, _inferno.createComponentVNode)(2, _virtual_scrolling.VirtualScrolling, {
        mode: virtualScrollingMode
    })], 0)
};
exports.viewFunction = viewFunction;
var DataGridNextScrollingProps = {
    mode: "standard"
};
exports.DataGridNextScrollingProps = DataGridNextScrollingProps;
var DataGridNextScrolling = function(_BaseInfernoComponent) {
    _inheritsLoose(DataGridNextScrolling, _BaseInfernoComponent);

    function DataGridNextScrolling(props) {
        var _this;
        _this = _BaseInfernoComponent.call(this, props) || this;
        _this.state = {};
        return _this
    }
    var _proto = DataGridNextScrolling.prototype;
    _proto.render = function() {
        var props = this.props;
        return viewFunction({
            props: _extends({}, props),
            isVirtualScrolling: this.isVirtualScrolling,
            virtualScrollingMode: this.virtualScrollingMode,
            restAttributes: this.restAttributes
        })
    };
    _createClass(DataGridNextScrolling, [{
        key: "isVirtualScrolling",
        get: function() {
            var mode = this.props.mode;
            return "standard" !== mode
        }
    }, {
        key: "virtualScrollingMode",
        get: function() {
            return "virtual" === this.props.mode ? "virtual" : "infinite"
        }
    }, {
        key: "restAttributes",
        get: function() {
            var _this$props = this.props,
                restProps = (_this$props.mode, _objectWithoutProperties(_this$props, _excluded));
            return restProps
        }
    }]);
    return DataGridNextScrolling
}(_inferno2.BaseInfernoComponent);
exports.DataGridNextScrolling = DataGridNextScrolling;
DataGridNextScrolling.defaultProps = DataGridNextScrollingProps;
