// devextreme/cjs/renovation/ui/grids/data_grid_next/widgets/data_row.js
/**
 * DevExtreme (cjs/renovation/ui/grids/data_grid_next/widgets/data_row.js)
 * Version: 22.1.6
 * Build date: Tue Oct 18 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.viewFunction = exports.DataRowProps = exports.DataRow = void 0;
var _inferno = require("inferno");
var _inferno2 = require("@devextreme/runtime/inferno");
var _context = require("../../../../utils/plugin/context");
var _data_cell = require("./data_cell");
var _row_base = require("./row_base");
var _classes = _interopRequireDefault(require("../classes"));
var _excluded = ["columns", "row", "rowIndex"];

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    }
}

function _objectWithoutProperties(source, excluded) {
    if (null == source) {
        return {}
    }
    var target = _objectWithoutPropertiesLoose(source, excluded);
    var key, i;
    if (Object.getOwnPropertySymbols) {
        var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
        for (i = 0; i < sourceSymbolKeys.length; i++) {
            key = sourceSymbolKeys[i];
            if (excluded.indexOf(key) >= 0) {
                continue
            }
            if (!Object.prototype.propertyIsEnumerable.call(source, key)) {
                continue
            }
            target[key] = source[key]
        }
    }
    return target
}

function _objectWithoutPropertiesLoose(source, excluded) {
    if (null == source) {
        return {}
    }
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;
    for (i = 0; i < sourceKeys.length; i++) {
        key = sourceKeys[i];
        if (excluded.indexOf(key) >= 0) {
            continue
        }
        target[key] = source[key]
    }
    return target
}

function _defineProperty(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        })
    } else {
        obj[key] = value
    }
    return obj
}

function _extends() {
    _extends = Object.assign ? Object.assign.bind() : function(target) {
        for (var i = 1; i < arguments.length; i++) {
            var source = arguments[i];
            for (var key in source) {
                if (Object.prototype.hasOwnProperty.call(source, key)) {
                    target[key] = source[key]
                }
            }
        }
        return target
    };
    return _extends.apply(this, arguments)
}

function _defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) {
            descriptor.writable = true
        }
        Object.defineProperty(target, descriptor.key, descriptor)
    }
}

function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) {
        _defineProperties(Constructor.prototype, protoProps)
    }
    if (staticProps) {
        _defineProperties(Constructor, staticProps)
    }
    Object.defineProperty(Constructor, "prototype", {
        writable: false
    });
    return Constructor
}

function _assertThisInitialized(self) {
    if (void 0 === self) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called")
    }
    return self
}

function _inheritsLoose(subClass, superClass) {
    subClass.prototype = Object.create(superClass.prototype);
    subClass.prototype.constructor = subClass;
    _setPrototypeOf(subClass, superClass)
}

function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(o, p) {
        o.__proto__ = p;
        return o
    };
    return _setPrototypeOf(o, p)
}
var viewFunction = function(viewModel) {
    var RowTemplate = viewModel.rowTemplate;
    var _viewModel$props = viewModel.props,
        columns = _viewModel$props.columns,
        row = _viewModel$props.row,
        rowIndex = _viewModel$props.rowIndex;
    return RowTemplate ? (0, _inferno.createComponentVNode)(2, RowTemplate, {
        row: row,
        rowIndex: rowIndex
    }) : (0, _inferno.createComponentVNode)(2, _row_base.RowBase, {
        row: row,
        children: columns.map((function(column, index) {
            return (0, _inferno.createComponentVNode)(2, _data_cell.DataCell, {
                columnIndex: index,
                countColumn: columns.length,
                column: column,
                cellTemplate: column.cellTemplate,
                cellContainerTemplate: column.cellContainerTemplate,
                row: row
            }, index)
        }))
    })
};
exports.viewFunction = viewFunction;
var DataRowProps = {
    row: Object.freeze({
        data: {},
        rowType: "data"
    }),
    rowIndex: 0,
    columns: Object.freeze([])
};
exports.DataRowProps = DataRowProps;
var DataRow = function(_InfernoComponent) {
    _inheritsLoose(DataRow, _InfernoComponent);

    function DataRow(props) {
        var _this;
        _this = _InfernoComponent.call(this, props) || this;
        _this.state = {};
        _this.extendDataRowClasses = _this.extendDataRowClasses.bind(_assertThisInitialized(_this));
        return _this
    }
    var _proto = DataRow.prototype;
    _proto.createEffects = function() {
        return [new _inferno2.InfernoEffect(this.extendDataRowClasses, [this.plugins])]
    };
    _proto.updateEffects = function() {
        var _this$_effects$;
        null === (_this$_effects$ = this._effects[0]) || void 0 === _this$_effects$ ? void 0 : _this$_effects$.update([this.plugins])
    };
    _proto.extendDataRowClasses = function() {
        return this.plugins.extend(_row_base.RowClassesGetter, 1, (function(base) {
            return function(row) {
                if ("data" === row.rowType) {
                    return _extends({}, base(row), _defineProperty({}, _classes.default.dataRow, true))
                }
                return base(row)
            }
        }))
    };
    _proto.render = function() {
        var props = this.props;
        return viewFunction({
            props: _extends({}, props),
            plugins: this.plugins,
            rowTemplate: this.rowTemplate,
            restAttributes: this.restAttributes
        })
    };
    _createClass(DataRow, [{
        key: "plugins",
        get: function() {
            if ("PluginsContext" in this.context) {
                return this.context.PluginsContext
            }
            return _context.PluginsContext
        }
    }, {
        key: "rowTemplate",
        get: function() {
            return this.props.row.template
        }
    }, {
        key: "restAttributes",
        get: function() {
            var _this$props = this.props,
                restProps = (_this$props.columns, _this$props.row, _this$props.rowIndex, _objectWithoutProperties(_this$props, _excluded));
            return restProps
        }
    }]);
    return DataRow
}(_inferno2.InfernoComponent);
exports.DataRow = DataRow;
DataRow.defaultProps = DataRowProps;
