// devextreme/cjs/renovation/utils/plugin/placeholder.js
/**
 * DevExtreme (cjs/renovation/utils/plugin/placeholder.js)
 * Version: 22.1.6
 * Build date: Tue Oct 18 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.viewFunction = exports.PlaceholderProps = exports.Placeholder = void 0;
var _inferno = require("inferno");
var _inferno2 = require("@devextreme/runtime/inferno");
var _placeholder_item = require("./placeholder_item");
var _context = require("./context");
var _excluded = ["children", "column", "type"];

function _objectWithoutProperties(source, excluded) {
    if (null == source) {
        return {}
    }
    var target = _objectWithoutPropertiesLoose(source, excluded);
    var key, i;
    if (Object.getOwnPropertySymbols) {
        var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
        for (i = 0; i < sourceSymbolKeys.length; i++) {
            key = sourceSymbolKeys[i];
            if (excluded.indexOf(key) >= 0) {
                continue
            }
            if (!Object.prototype.propertyIsEnumerable.call(source, key)) {
                continue
            }
            target[key] = source[key]
        }
    }
    return target
}

function _objectWithoutPropertiesLoose(source, excluded) {
    if (null == source) {
        return {}
    }
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;
    for (i = 0; i < sourceKeys.length; i++) {
        key = sourceKeys[i];
        if (excluded.indexOf(key) >= 0) {
            continue
        }
        target[key] = source[key]
    }
    return target
}

function _extends() {
    _extends = Object.assign ? Object.assign.bind() : function(target) {
        for (var i = 1; i < arguments.length; i++) {
            var source = arguments[i];
            for (var key in source) {
                if (Object.prototype.hasOwnProperty.call(source, key)) {
                    target[key] = source[key]
                }
            }
        }
        return target
    };
    return _extends.apply(this, arguments)
}

function _defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) {
            descriptor.writable = true
        }
        Object.defineProperty(target, descriptor.key, descriptor)
    }
}

function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) {
        _defineProperties(Constructor.prototype, protoProps)
    }
    if (staticProps) {
        _defineProperties(Constructor, staticProps)
    }
    Object.defineProperty(Constructor, "prototype", {
        writable: false
    });
    return Constructor
}

function _assertThisInitialized(self) {
    if (void 0 === self) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called")
    }
    return self
}

function _inheritsLoose(subClass, superClass) {
    subClass.prototype = Object.create(superClass.prototype);
    subClass.prototype.constructor = subClass;
    _setPrototypeOf(subClass, superClass)
}

function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(o, p) {
        o.__proto__ = p;
        return o
    };
    return _setPrototypeOf(o, p)
}
var viewFunction = function(_ref) {
    var componentDeps = _ref.componentDeps,
        componentTypes = _ref.componentTypes,
        _ref$props = _ref.props,
        children = _ref$props.children,
        column = _ref$props.column;
    return (0, _inferno.createComponentVNode)(2, _placeholder_item.PlaceholderItem, {
        componentTypes: componentTypes,
        componentDeps: componentDeps,
        column: column,
        index: 0,
        children: children
    })
};
exports.viewFunction = viewFunction;
var PlaceholderProps = {};
exports.PlaceholderProps = PlaceholderProps;
var Placeholder = function(_InfernoComponent) {
    _inheritsLoose(Placeholder, _InfernoComponent);

    function Placeholder(props) {
        var _this;
        _this = _InfernoComponent.call(this, props) || this;
        _this.state = {
            componentTypes: [],
            componentDeps: []
        };
        _this.updateComponentTypes = _this.updateComponentTypes.bind(_assertThisInitialized(_this));
        return _this
    }
    var _proto = Placeholder.prototype;
    _proto.createEffects = function() {
        return [new _inferno2.InfernoEffect(this.updateComponentTypes, [this.plugins, this.props.type])]
    };
    _proto.updateEffects = function() {
        var _this$_effects$;
        null === (_this$_effects$ = this._effects[0]) || void 0 === _this$_effects$ ? void 0 : _this$_effects$.update([this.plugins, this.props.type])
    };
    _proto.updateComponentTypes = function() {
        var _this2 = this;
        return this.plugins.watch(this.props.type, (function(items) {
            _this2.setState((function(__state_argument) {
                return {
                    componentTypes: items.map((function(item) {
                        return item.component
                    })).reverse()
                }
            }));
            _this2.setState((function(__state_argument) {
                return {
                    componentDeps: items.map((function(item) {
                        return item.deps
                    })).reverse()
                }
            }))
        }))
    };
    _proto.render = function() {
        var props = this.props;
        return viewFunction({
            props: _extends({}, props),
            componentTypes: this.state.componentTypes,
            componentDeps: this.state.componentDeps,
            plugins: this.plugins,
            restAttributes: this.restAttributes
        })
    };
    _createClass(Placeholder, [{
        key: "plugins",
        get: function() {
            if ("PluginsContext" in this.context) {
                return this.context.PluginsContext
            }
            return _context.PluginsContext
        }
    }, {
        key: "restAttributes",
        get: function() {
            var _this$props = this.props,
                restProps = (_this$props.children, _this$props.column, _this$props.type, _objectWithoutProperties(_this$props, _excluded));
            return restProps
        }
    }]);
    return Placeholder
}(_inferno2.InfernoComponent);
exports.Placeholder = Placeholder;
Placeholder.defaultProps = PlaceholderProps;
