// devextreme/esm/renovation/utils/plugin/value_setter.js
/**
 * DevExtreme (esm/renovation/utils/plugin/value_setter.js)
 * Version: 22.1.6
 * Build date: Tue Oct 18 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
var _excluded = ["type", "value"];
import {
    createVNode
} from "inferno";
import {
    InfernoEffect,
    InfernoComponent
} from "@devextreme/runtime/inferno";
import {
    PluginsContext
} from "./context";
export var viewFunction = () => createVNode(1, "div");
export var ValueSetterProps = {};
export class ValueSetter extends InfernoComponent {
    constructor(props) {
        super(props);
        this.state = {};
        this.updatePluginValue = this.updatePluginValue.bind(this)
    }
    get plugins() {
        if ("PluginsContext" in this.context) {
            return this.context.PluginsContext
        }
        return PluginsContext
    }
    createEffects() {
        return [new InfernoEffect(this.updatePluginValue, [this.plugins, this.props.type, this.props.value])]
    }
    updateEffects() {
        var _this$_effects$;
        null === (_this$_effects$ = this._effects[0]) || void 0 === _this$_effects$ ? void 0 : _this$_effects$.update([this.plugins, this.props.type, this.props.value])
    }
    updatePluginValue() {
        this.plugins.set(this.props.type, this.props.value)
    }
    get restAttributes() {
        var _this$props = this.props,
            restProps = _objectWithoutPropertiesLoose(_this$props, _excluded);
        return restProps
    }
    render() {
        this.props;
        return viewFunction()
    }
}
ValueSetter.defaultProps = ValueSetterProps;
