// devextreme/cjs/renovation/ui/grids/data_grid_next/data_grid_next_full.j.js
/**
 * DevExtreme (cjs/renovation/ui/grids/data_grid_next/data_grid_next_full.j.js)
 * Version: 22.1.6
 * Build date: Tue Oct 18 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.default = void 0;
var _component_registrator = _interopRequireDefault(require("../../../../core/component_registrator"));
var _component = _interopRequireDefault(require("../../../component_wrapper/common/component"));
var _data_grid_next_full = require("./data_grid_next_full");

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    }
}

function _defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) {
            descriptor.writable = true
        }
        Object.defineProperty(target, descriptor.key, descriptor)
    }
}

function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) {
        _defineProperties(Constructor.prototype, protoProps)
    }
    if (staticProps) {
        _defineProperties(Constructor, staticProps)
    }
    Object.defineProperty(Constructor, "prototype", {
        writable: false
    });
    return Constructor
}

function _inheritsLoose(subClass, superClass) {
    subClass.prototype = Object.create(superClass.prototype);
    subClass.prototype.constructor = subClass;
    _setPrototypeOf(subClass, superClass)
}

function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(o, p) {
        o.__proto__ = p;
        return o
    };
    return _setPrototypeOf(o, p)
}
var DataGridNext = function(_BaseComponent) {
    _inheritsLoose(DataGridNext, _BaseComponent);

    function DataGridNext() {
        return _BaseComponent.apply(this, arguments) || this
    }
    var _proto = DataGridNext.prototype;
    _proto.getProps = function() {
        var props = _BaseComponent.prototype.getProps.call(this);
        props.onKeyDown = this._wrapKeyDownHandler(props.onKeyDown);
        return props
    };
    _proto.refresh = function() {
        var _this$viewRef;
        return null === (_this$viewRef = this.viewRef) || void 0 === _this$viewRef ? void 0 : _this$viewRef.refresh.apply(_this$viewRef, arguments)
    };
    _proto._getActionConfigs = function() {
        return {
            onDataErrorOccurred: {},
            onClick: {}
        }
    };
    _createClass(DataGridNext, [{
        key: "_propsInfo",
        get: function() {
            return {
                twoWay: [
                    ["dataState", "defaultDataState", "dataStateChange"]
                ],
                allowNull: [],
                elements: [],
                templates: ["noDataTemplate"],
                props: ["paging", "pager", "selection", "masterDetail", "dataSource", "remoteOperations", "cacheEnabled", "keyExpr", "columns", "onDataErrorOccurred", "noDataTemplate", "defaultDataState", "dataStateChange", "className", "accessKey", "activeStateEnabled", "disabled", "focusStateEnabled", "height", "hint", "hoverStateEnabled", "onClick", "onKeyDown", "rtlEnabled", "tabIndex", "visible", "width", "dataState"]
            }
        }
    }, {
        key: "_viewComponent",
        get: function() {
            return _data_grid_next_full.DataGridNext
        }
    }]);
    return DataGridNext
}(_component.default);
exports.default = DataGridNext;
(0, _component_registrator.default)("dxDataGridNext", DataGridNext);
DataGridNext.defaultOptions = _data_grid_next_full.defaultOptions;
module.exports = exports.default;
module.exports.default = exports.default;
