// devextreme/renovation/utils/plugin/placeholder_item_renderer.js
/**
 * DevExtreme (renovation/utils/plugin/placeholder_item_renderer.js)
 * Version: 22.1.6
 * Build date: Tue Oct 18 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.viewFunction = exports.PlaceholderItemRendererProps = exports.PlaceholderItemRenderer = void 0;
var _inferno = require("inferno");
var _inferno2 = require("@devextreme/runtime/inferno");
var _excluded = ["baseTemplate", "column", "currentTemplate", "deps"];

function _objectWithoutProperties(source, excluded) {
    if (null == source) {
        return {}
    }
    var target = _objectWithoutPropertiesLoose(source, excluded);
    var key, i;
    if (Object.getOwnPropertySymbols) {
        var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
        for (i = 0; i < sourceSymbolKeys.length; i++) {
            key = sourceSymbolKeys[i];
            if (excluded.indexOf(key) >= 0) {
                continue
            }
            if (!Object.prototype.propertyIsEnumerable.call(source, key)) {
                continue
            }
            target[key] = source[key]
        }
    }
    return target
}

function _objectWithoutPropertiesLoose(source, excluded) {
    if (null == source) {
        return {}
    }
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;
    for (i = 0; i < sourceKeys.length; i++) {
        key = sourceKeys[i];
        if (excluded.indexOf(key) >= 0) {
            continue
        }
        target[key] = source[key]
    }
    return target
}

function _defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) {
            descriptor.writable = true
        }
        Object.defineProperty(target, descriptor.key, descriptor)
    }
}

function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) {
        _defineProperties(Constructor.prototype, protoProps)
    }
    if (staticProps) {
        _defineProperties(Constructor, staticProps)
    }
    Object.defineProperty(Constructor, "prototype", {
        writable: false
    });
    return Constructor
}

function _inheritsLoose(subClass, superClass) {
    subClass.prototype = Object.create(superClass.prototype);
    subClass.prototype.constructor = subClass;
    _setPrototypeOf(subClass, superClass)
}

function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(o, p) {
        o.__proto__ = p;
        return o
    };
    return _setPrototypeOf(o, p)
}

function _extends() {
    _extends = Object.assign ? Object.assign.bind() : function(target) {
        for (var i = 1; i < arguments.length; i++) {
            var source = arguments[i];
            for (var key in source) {
                if (Object.prototype.hasOwnProperty.call(source, key)) {
                    target[key] = source[key]
                }
            }
        }
        return target
    };
    return _extends.apply(this, arguments)
}
var viewFunction = function(_ref) {
    var _ref$props = _ref.props,
        baseTemplate = _ref$props.baseTemplate,
        column = _ref$props.column,
        CurrentTemplate = _ref$props.currentTemplate,
        deps = _ref$props.deps;
    return CurrentTemplate({
        deps: deps,
        column: column,
        baseTemplate: baseTemplate
    })
};
exports.viewFunction = viewFunction;
var PlaceholderItemRendererProps = {
    deps: Object.freeze([])
};
exports.PlaceholderItemRendererProps = PlaceholderItemRendererProps;
var getTemplate = function(TemplateProp) {
    return TemplateProp && (TemplateProp.defaultProps ? function(props) {
        return (0, _inferno.normalizeProps)((0, _inferno.createComponentVNode)(2, TemplateProp, _extends({}, props)))
    } : TemplateProp)
};
var PlaceholderItemRenderer = function(_BaseInfernoComponent) {
    _inheritsLoose(PlaceholderItemRenderer, _BaseInfernoComponent);

    function PlaceholderItemRenderer(props) {
        var _this;
        _this = _BaseInfernoComponent.call(this, props) || this;
        _this.state = {};
        return _this
    }
    var _proto = PlaceholderItemRenderer.prototype;
    _proto.render = function() {
        var props = this.props;
        return viewFunction({
            props: _extends({}, props, {
                currentTemplate: getTemplate(props.currentTemplate),
                baseTemplate: getTemplate(props.baseTemplate)
            }),
            restAttributes: this.restAttributes
        })
    };
    _createClass(PlaceholderItemRenderer, [{
        key: "restAttributes",
        get: function() {
            var _this$props = this.props,
                restProps = (_this$props.baseTemplate, _this$props.column, _this$props.currentTemplate, _this$props.deps, _objectWithoutProperties(_this$props, _excluded));
            return restProps
        }
    }]);
    return PlaceholderItemRenderer
}(_inferno2.BaseInfernoComponent);
exports.PlaceholderItemRenderer = PlaceholderItemRenderer;
PlaceholderItemRenderer.defaultProps = PlaceholderItemRendererProps;
