// devextreme/esm/renovation/ui/grids/data_grid_next/data_grid_next_full.j.js
/**
 * DevExtreme (esm/renovation/ui/grids/data_grid_next/data_grid_next_full.j.js)
 * Version: 22.1.6
 * Build date: Tue Oct 18 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import registerComponent from "../../../../core/component_registrator";
import BaseComponent from "../../../component_wrapper/common/component";
import {
    DataGridNext as DataGridNextComponent,
    defaultOptions
} from "./data_grid_next_full";
export default class DataGridNext extends BaseComponent {
    getProps() {
        var props = super.getProps();
        props.onKeyDown = this._wrapKeyDownHandler(props.onKeyDown);
        return props
    }
    refresh() {
        var _this$viewRef;
        return null === (_this$viewRef = this.viewRef) || void 0 === _this$viewRef ? void 0 : _this$viewRef.refresh(...arguments)
    }
    _getActionConfigs() {
        return {
            onDataErrorOccurred: {},
            onClick: {}
        }
    }
    get _propsInfo() {
        return {
            twoWay: [
                ["dataState", "defaultDataState", "dataStateChange"]
            ],
            allowNull: [],
            elements: [],
            templates: ["noDataTemplate"],
            props: ["paging", "pager", "selection", "masterDetail", "dataSource", "remoteOperations", "cacheEnabled", "keyExpr", "columns", "onDataErrorOccurred", "noDataTemplate", "defaultDataState", "dataStateChange", "className", "accessKey", "activeStateEnabled", "disabled", "focusStateEnabled", "height", "hint", "hoverStateEnabled", "onClick", "onKeyDown", "rtlEnabled", "tabIndex", "visible", "width", "dataState"]
        }
    }
    get _viewComponent() {
        return DataGridNextComponent
    }
}
registerComponent("dxDataGridNext", DataGridNext);
DataGridNext.defaultOptions = defaultOptions;
