// devextreme/renovation/ui/grids/data_grid_next/classes.js
/**
 * DevExtreme (renovation/ui/grids/data_grid_next/classes.js)
 * Version: 22.1.6
 * Build date: Tue Oct 18 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.default = void 0;
var _default = {
    dataGrid: "dx-datagrid",
    gridBaseContainer: "dx-gridbase-container",
    rowsView: "dx-datagrid-rowsview",
    noWrap: "dx-datagrid-nowrap",
    content: "dx-datagrid-content",
    headers: "dx-datagrid-headers",
    afterHeaders: "dx-datagrid-after-headers",
    scrollContainer: "dx-datagrid-scroll-container",
    noData: "dx-datagrid-nodata",
    empty: "dx-empty",
    firstChild: "dx-first-child",
    lastChild: "dx-last-child",
    row: "dx-row",
    dataRow: "dx-data-row",
    headerRow: "dx-header-row",
    masterDetailRow: "dx-master-detail-row",
    virtualRow: "dx-virtual-row",
    pager: "dx-datagrid-pager",
    checkboxSize: "dx-datagrid-checkbox-size",
    selectCheckbox: "dx-select-checkbox",
    selectedRow: "dx-selection",
    columnLines: "dx-column-lines",
    textContent: "dx-datagrid-text-content",
    textContentAlignmentLeft: "dx-text-content-alignment-left",
    action: "dx-datagrid-action",
    cellFocusDisabled: "dx-cell-focus-disabled",
    groupSpace: "dx-datagrid-group-space",
    table: "dx-datagrid-table",
    fixedTable: "dx-datagrid-table-fixed",
    commandExpand: "dx-command-expand",
    expand: "dx-datagrid-expand",
    groupOpened: "dx-datagrid-group-opened",
    groupClosed: "dx-datagrid-group-closed"
};
exports.default = _default;
module.exports = exports.default;
module.exports.default = exports.default;
