// devextreme/esm/renovation/ui/grids/data_grid_next/widgets/no_data_text.js
/**
 * DevExtreme (esm/renovation/ui/grids/data_grid_next/widgets/no_data_text.js)
 * Version: 22.1.6
 * Build date: Tue Oct 18 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import _extends from "@babel/runtime/helpers/esm/extends";
var _excluded = ["template"];
import {
    createVNode,
    createComponentVNode,
    normalizeProps
} from "inferno";
import {
    BaseInfernoComponent
} from "@devextreme/runtime/inferno";
import messageLocalization from "../../../../../localization/message";
import CLASSES from "../classes";
export var viewFunction = viewModel => createVNode(1, "div", CLASSES.noData, [viewModel.props.template && viewModel.props.template({}), !viewModel.props.template && viewModel.text], 0);
export var NoDataTextProps = {};
var getTemplate = TemplateProp => TemplateProp && (TemplateProp.defaultProps ? props => normalizeProps(createComponentVNode(2, TemplateProp, _extends({}, props))) : TemplateProp);
export class NoDataText extends BaseInfernoComponent {
    constructor(props) {
        super(props);
        this.state = {}
    }
    get text() {
        return messageLocalization.format("dxDataGrid-noDataText")
    }
    get restAttributes() {
        var _this$props = this.props,
            restProps = _objectWithoutPropertiesLoose(_this$props, _excluded);
        return restProps
    }
    render() {
        var props = this.props;
        return viewFunction({
            props: _extends({}, props, {
                template: getTemplate(props.template)
            }),
            text: this.text,
            restAttributes: this.restAttributes
        })
    }
}
NoDataText.defaultProps = NoDataTextProps;
