// devextreme/esm/renovation/ui/grids/data_grid_next/widgets/table.js
/**
 * DevExtreme (esm/renovation/ui/grids/data_grid_next/widgets/table.js)
 * Version: 22.1.6
 * Build date: Tue Oct 18 2022
 *
 * Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
var _excluded = ["children"];
import {
    createVNode
} from "inferno";
import {
    BaseInfernoComponent
} from "@devextreme/runtime/inferno";
import CLASSES from "../classes";
export var viewFunction = viewModel => createVNode(1, "table", "".concat(CLASSES.table, " ").concat(CLASSES.fixedTable), createVNode(1, "tbody", null, viewModel.props.children, 0, {
    role: "presentation"
}), 2, {
    role: "presentation"
});
export var TableProps = {};
export class Table extends BaseInfernoComponent {
    constructor(props) {
        super(props);
        this.state = {}
    }
    get restAttributes() {
        var _this$props = this.props,
            restProps = _objectWithoutPropertiesLoose(_this$props, _excluded);
        return restProps
    }
    render() {
        var props = this.props;
        return viewFunction({
            props: _extends({}, props),
            restAttributes: this.restAttributes
        })
    }
}
Table.defaultProps = TableProps;
